File: 0768-max-chunks-to-make-sorted-ii/0768-max-chunks-to-make-sorted-ii.cpp
class Solution {
public:
    int maxChunksToSorted(vector<int>& arr) {
        int count=0;
        int n=arr.size();
        int mini=INT_MAX;

        vector<int>mn_till(n,0);

        for(int i=n-1;i>=0;i--){
            mini=min(mini,arr[i]);
            mn_till[i]=mini;
        }

        int maxi=arr[0];

        for(int i=1;i<n;i++){
            if(maxi<=mn_till[i]){
                count++;
                maxi=arr[i];
            }
            else{
                maxi=max(maxi,arr[i]);
            }
        }
        return count+1;
    }
};

File: 0768-max-chunks-to-make-sorted-ii/0768-max-chunks-to-make-sorted-ii.java
class Solution {
    public int maxChunksToSorted(int[] input) {
        int n = input.length;
        int[] suffixR = new int[n];

        suffixR[n - 1] = input[n - 1];

        for (int i = n - 2; i >= 0; i--) {
            suffixR[i] = Math.min(suffixR[i + 1], input[i]);
        }

        int leftMax = input[0];
        int chunks = 0;

        for (int i = 1; i < n; i++) {
            if (leftMax <= suffixR[i]) {
                chunks++;
                leftMax = input[i];   // important reset
            } else {
                leftMax = Math.max(leftMax, input[i]);
            }
        }

        return chunks + 1;
    }
}